src/lua/datetime.lua

```diff
@@ -0,0 +1,623 @@
+end
+
+local parse_zone
+
+local function datetime_new_raw(epoch, nsec, tzoffset, tzindex)
+    local dt_obj = ffi.new(datetime_t)
+    dt_obj.epoch = epoch
+    dt_obj.nsec = nsec
+    dt_obj.tzoffset = tzoffset or 0
+    dt_obj.tzindex = tzindex or 0
+    return dt_obj
+end
+
+local function datetime_new_dt(dt, secs, fraction, offset)
+    local epochV = dt ~= nil and (builtin.tnt_dt_rdn(dt) - DT_EPOCH_1970_OFFSET) *
+                   SECS_PER_DAY or 0
+    local secsV = secs or 0
+    local fracV = fraction or 0
+    local ofsV = offset or 0
+    return datetime_new_raw(epochV + secsV - ofsV * 60, fracV, ofsV)
+end
+
+local function get_timezone(offset)
+    if type(offset) == 'number' then
+        return offset
+    elseif type(offset) == 'string' then
+        return parse_zone(offset)
+    end
+end
+
+-- create datetime given attribute values from obj
+local function datetime_new(obj)
+    if obj == nil or type(obj) ~= 'table' then
+        return datetime_new_raw(0, 0, 0)
+    end
+
+    local ymd = false
+    local hms = false
+    local dt = DT_EPOCH_1970_OFFSET -- default is 1970-01-01
+
+    local y = obj.year
+    if y ~= nil then
+        check_range(y, {1, 9999}, 'year')
+        ymd = true
+    end
+    local M = obj.month
+    if M ~= nil then
+        check_range(M, {1, 12}, 'month')
+        ymd = true
+    end
+    local d = obj.day
+    if d ~= nil then
+        check_range(d, {1, 31, -1}, 'day')
+        ymd = true
+    end
+    local h = obj.hour
+    if h ~= nil then
+        check_range(h, {0, 23}, 'hour')
+        hms = true
+    end
+    local m = obj.min
+    if m ~= nil then
+        check_range(m, {0, 59}, 'min')
+        hms = true
+    end
+    local s = obj.sec
+    if s ~= nil then
+        check_range(s, {0, 60}, 'sec')
+        hms = true
+    end
+    local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
+    -- if there are separate nsec, usec, or msec provided then
+    -- timestamp should be integer
+    local int_ts = nsec ~= nil or usec ~= nil or msec ~= nil
+
+    local ts = obj.timestamp
+    local fraction
+    if ts ~= nil then
+        s, fraction = math_modf(ts)
+        if not int_ts then
+            nsec = fraction * 1e9
+        end
+        hms = true
+    end
+
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset)
+        check_range(offset, {-720, 720}, offset)
+    end
+
+    if obj.tz ~= nil then
+        nyi('tz')
+    end
+
+    -- .year, .month, .day
+    if ymd then
+        y = y or 1970
+        M = M or 1
+        if d ~= nil and d < 0 then
+            d = builtin.tnt_dt_days_in_month(y, M)
+        end
+        dt = builtin.tnt_dt_from_ymd(y, M, d)
+    end
+
+    -- .hour, .minute, .second
+    local secs = 0
+    if hms then
+        secs = (h or 0) * 3600 + (m or 0) * 60 + (s or 0)
+    end
+
+    return datetime_new_dt(dt, secs, nsec, offset or 0)
+end
+
+local sz = 48
+local buff = ffi.new('char[?]', sz)
```

Comment:
By having a global buffer you added exactly the same bug I warned about in the previous review. If any of the functions using the buffer will be called from a GC handler, you will turn it into garbage. For big global shared objects there is `buffer.ffi_stash_new`. But TBH, I see there more serious issues than this one, so it is up to you. Must be either created on each call, or use `ffi_stash_new` if want.

Reply:
You know, after your review there Vova was here, thus we came up with global buffers 🤷‍♀️

For a moment, have returned local variables usage:
```diff
@@ -96,7 +96,7 @@ local NANOS_PER_SEC    = 1000000000
 -- Unix, January 1, 1970, Thursday
 local DAYS_EPOCH_OFFSET = 719163
 local SECS_EPOCH_OFFSET = DAYS_EPOCH_OFFSET * SECS_PER_DAY
-
+local TOSTRING_BUFF_SZ = 48
 
 local datetime_t = ffi.typeof('struct datetime')
 
@@ -370,9 +370,6 @@ local function datetime_new(obj)
     return datetime_new_dt(dt, secs, nsec, offset or 0)
 end
 
-local sz = 48
-local buff = ffi.new('char[?]', sz)
-
 --[[
     Convert to text datetime values
 
@@ -383,8 +380,9 @@ local buff = ffi.new('char[?]', sz)
 local function datetime_tostring(self)
     check_date(self, 'datetime.tostring()')
 
-    local len = builtin.datetime_to_string(buff, sz, self)
-    assert(len < sz)
+    local buff = ffi.new('char[?]', TOSTRING_BUFF_SZ)
+    local len = builtin.datetime_to_string(buff, TOSTRING_BUFF_SZ, self)
+    assert(len < TOSTRING_BUFF_SZ)
     return ffi.string(buff)
 end
 
```

I'm confused with your recommendation to use `buffer.ffi_stash_new` because I believed that stash is used when we take/put frequently same size allocated structures. There is no such pattern used in our code for `date:tostring()`. How did you assume it will be used?

Reply:
Yes, it is used for frequent put/take same-size allocations. This is the case here. You use a temporary string buffer to call a C-function and then copy it into Lua memory.